packages/voter-dapp/src/RetrieveRewards.js

```diff
@@ -204,16 +212,28 @@ function RetrieveRewards({ votingAccount }) {
 
   // Construct the claim rewards transaction.
   const rewardsTxn = useRetrieveRewardsTxn(retrievedRewardsEvents, reveals, votingAccount);
+  if (rewardsTxn.pendingRewardsPromise) {
```

Comment:
Just as a side note, im noticing now how the reward state gets in here: The `useRetreiveRewards` is just a regular function that is going to get called each time this component renders.  This does not seem right to me, it should be wrapped in useEffect or useMemo which sets state asynchronously and is more controlled on when to update. 

This has some implications on your implementation but I think as long as it works, its ok. If there are a lot of re-renders though it could be a problem.

Reply:
Great question. I just tried putting `useRetrieveRewards` inside a `useEffect()` hook and received an error that you cannot place Hooks inside of a callback, indicating that `useRetrieveRewardsTxn` is itself a hook. I think its a hook because it returns return values from a `useCacheSend` hook.

Reply:
Agreed, I think there may be ways to memo-ize some of the event processing logic inside this hook (without memoizing the hook itself). Drizzle ensures that the blockchain calls, themselves, are cached and not called more often than necessary. My guess is that memoization of the processing logic would help, but due to the infrequency of votes (which means the event data shouldn’t be huge), I would guess that it wouldn’t make a huge diff in perf. I think it would be worth investing the time here if this dapp weren’t in maintenance mode, but since we’re actively working on a rewrite, it doesn’t seem super worthwhile.

Reply:
@daywiss are you saying we should memo-ize the promise that’s returned from the hook and then mono-ize this function call whenever that promise reference changes?

Reply:
Wait, I think I see what you’re saying. As it’s currently written, it will create a slow-moving loop. It calls the promise. The promise sets state, which triggers a new promise. Any render not triggered by a previous promise will create a new infinite chain of promises, so this could build up over time.

Reply:
I do think this could cause a real problem.

Reply:
We could remove `pendingRewards` from the state so that the promise resolving does not trigger a re-render?

Reply:
But then how would it be reflected on-screen?

Reply:
You could add something like the following to `useRetrieveRewardsTxn` (and pass in a callback function that would set the state):

```
useEffect(() => {
VotingContract.methods.retrieveRewards(votingAccount, oldestUnclaimedRound.toString(), toRetrieve)
      .call().then(callback);
}, [votingAccount, oldestUnclaimedRound.toString(), JSON.stringify(toRetrieve)]);
```

Reply:
I ended up adding a `useEffect` hook to the `RetrieveRewards` main render function, and return the params neccessary to call `retrieveRewards` from the `useRetrieveRewardsTxn` hook. This is because you cannot call `useEffect` conditionally (i.e. within an if statement) so it would have been difficult to call this within `useRetrieveRewardsTxn`.

Reply:
I think it would take some thought for a refactor on fixing this, it should probably be pushed to a new low priority issue and addressed if this is actually a problem imo

Reply:
@nicholaspai ^^Looks great!